atomic_reactor/plugins/pre_koji_delegate.py

```diff
@@ -0,0 +1,124 @@
+"""
+Copyright (c) 2019 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import unicode_literals, absolute_import
+
+import os
+import json
+import koji
+import time
+
+from atomic_reactor.plugin import PreBuildPlugin
+from atomic_reactor.plugins.pre_reactor_config import (get_koji_session, get_koji, NO_FALLBACK,
+                                                       get_openshift_session)
+from atomic_reactor.plugins.pre_check_and_set_rebuild import is_rebuild
+from atomic_reactor.constants import PLUGIN_KOJI_DELEGATE_KEY
+from atomic_reactor.util import get_build_json
+
+
+class KojiDelegatePlugin(PreBuildPlugin):
+    """
+    When autorebuild is enabled, delegate feature is enabled and
+    triggered_after_koji_task is not provided,
+    will submit new koji task for autorebuild, and cancel current build.
+    In all other cases plugin won't do anything
+    """
+
+    key = PLUGIN_KOJI_DELEGATE_KEY
+    is_allowed_to_fail = False  # We really want to stop the process
+
+    def __init__(self, tasker, workflow, triggered_after_koji_task=None):
+        """
+        constructor
+
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param triggered_after_koji_task: int, original koji task for autorebuild,
+            provided only when this plugin creates new koji task for autorebuild
+        """
+        # call parent constructor
+        super(KojiDelegatePlugin, self).__init__(tasker, workflow)
+
+        koji_setting = get_koji(self.workflow, NO_FALLBACK)
+        self.delegate_enabled = koji_setting.get('delegate_task', True)
+        self.task_priority = koji_setting.get('delegated_task_priority', None)
+        self.triggered_after_koji_task = triggered_after_koji_task
+        self.metadata = get_build_json().get("metadata", {})
+        self.kojisession = get_koji_session(self.workflow, NO_FALLBACK)
+        self.osbs = None
+
+    def cancel_build(self):
+        """
+        cancel current build
+        """
+        build_name = self.metadata.get("name")
+        if build_name:
+            self.osbs.cancel_build(build_name)
+
+    def delegate_task(self):
+        """
+        create new koji task for autorebuild
+        """
+        user_params = os.environ['USER_PARAMS']
+        user_data = json.loads(user_params)
+
+        git_uri = user_data.get('git_uri')
+        git_ref = user_data.get('git_ref')
+        source = "%s#%s" % (git_uri, git_ref)
+        container_target = user_data.get('koji_target')
+
+        koji_task_id = self.metadata.get('labels', {}).get('original-koji-task-id')
+        if not koji_task_id:
+            koji_task_id = self.metadata.get('labels', {}).get('koji-task-id')
+
+        task_opts = {}
+        for key in ('yum_repourls', 'git_branch', 'signing_intent', 'compose_ids'):
```

Comment:
This redefinition of `task_opts` seems prone to errors to me. How we make sure when tasks options are extended we will pass all of them here?

Reply:
well we are recreating task_opts based on user_params, which is basically like manually submitting build, so if tasks options will be extended and used also for autorebuild we will have to extend them here